compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -7134,25 +7136,77 @@ private void createTypeCastExprForRelationalExpr(BLangBinaryExpr binaryExpr, int
                                                                 int rhsExprTypeTag) {
         boolean isLhsIntegerType = TypeTags.isIntegerTypeTag(lhsExprTypeTag);
         boolean isRhsIntegerType = TypeTags.isIntegerTypeTag(rhsExprTypeTag);
+        BType lhsExprType = binaryExpr.lhsExpr.getBType();
+        BType rhsExprType = binaryExpr.rhsExpr.getBType();
 
         if ((isLhsIntegerType && isRhsIntegerType) || (lhsExprTypeTag == TypeTags.BYTE &&
                 rhsExprTypeTag == TypeTags.BYTE)) {
             return;
         }
 
+        if (lhsExprTypeTag == TypeTags.DECIMAL) {
+            if (rhsExprTypeTag == TypeTags.UNION && rhsExprType.isNullable()) {
+                binaryExpr.lhsExpr = addNilType(lhsExprType, binaryExpr.lhsExpr);
+                return;
+            }
+            binaryExpr.rhsExpr = createTypeCastExpr(binaryExpr.rhsExpr, binaryExpr.lhsExpr.getBType());
+            return;
+        }
+
+        if (rhsExprTypeTag == TypeTags.DECIMAL) {
+            if (lhsExprTypeTag == TypeTags.UNION && lhsExprType.isNullable()) {
+                binaryExpr.rhsExpr = addNilType(lhsExprType, binaryExpr.rhsExpr);
+                return;
+            }
+            binaryExpr.lhsExpr = createTypeCastExpr(binaryExpr.lhsExpr, binaryExpr.rhsExpr.getBType());
+            return;
+        }
+
+        if (lhsExprTypeTag == TypeTags.FLOAT) {
+            if (rhsExprTypeTag == TypeTags.UNION && rhsExprType.isNullable()) {
+                binaryExpr.lhsExpr = addNilType(lhsExprType, binaryExpr.lhsExpr);
+                return;
+            }
+            binaryExpr.rhsExpr = createTypeCastExpr(binaryExpr.rhsExpr, binaryExpr.lhsExpr.getBType());
+            return;
+        }
+
+        if (rhsExprTypeTag == TypeTags.FLOAT) {
+            if (lhsExprTypeTag == TypeTags.UNION && lhsExprType.isNullable()) {
+                binaryExpr.rhsExpr = addNilType(lhsExprType, binaryExpr.rhsExpr);
+                return;
+            }
+            binaryExpr.lhsExpr = createTypeCastExpr(binaryExpr.lhsExpr, binaryExpr.rhsExpr.getBType());
+            return;
+        }
```

Comment:
It seems these 4 `if block bodies` doing the same thing. Can we move the functionalities inside if block into a separate method?

Reply:
Addressed in https://github.com/ballerina-platform/ballerina-lang/pull/36191/commits/d45488667b6d5924ca4b8f957b0c559b0b80de7d.